### DISCUSSION TITLE
GUI Tools for RayOptics / OpticExplorer

### DISCUSSION THREAD

**Discussion by amegahed** [mjhoptics/ray-optics#122] [Ideas]
Hi.  At one point in a previous discussion, you stated: 

"My current interest is in GUI tools for first order optical design. This is an area that hasn't been well explored, IMHO, and is an area I've long been interested in."

I've been working on a similar project called OpticExplorer (www.opticexplorer.com) which has a nice interactive GUI but is lacking in analytical capabilities.   Here's an overview video:  
https://opticexplorer.sharedigm.com/services/public/api/video?path=%2FShared%2FVideos%2FWelcome%20To%20OpticExplorer.mp4

I don't know if there's a way that we could work together.   You can currently export .zmx files from OpticExplorer into RayOptics for analysis so that could provide RayOptics users with a loosely integrated form of GUI.    OpticExplorer could potentially also export ".roa" files as well.    

I frequently see a desire to have optical modeling and analysis tools that are more accessible than the commercial offerings but it's difficult for us as individual developers or small groups to take on this task.  I find myself wondering if it might be more achievable through some sort of collaboration.  

Cheers,

-abe.

OpticExplorer, http://www.opticexplorer.com

**Comment by mjhoptics** [comment]
Hello Abe,

First, let me congratulate you on Optic Explorer. It is a beautifully rendered app, well organized, and very responsive to interactive actions. One has to work hard to make even a mediocre GUI; a well executed one such as Optic Explorer deserves kudos!

As far as collaboration goes, I am happy to contribute to open source efforts to improve and widen the availability and capability of optical design tools. A couple of years ago, Bill Claff approached me with a diffractive optic modeling problem and we spent some time working thru examples and ultimately improved both of our codes. `ray-optics` can read his website's lens format. I would prefer reading your format directly, rather than going through .zmx, when building a bridge between programs. Alternatively, more directly thinking in terms of an API between a front end and back end might enable tighter integration between different code bases.

Thanks for reaching out and great work.
Mike

**Reply by amegahed** [reply]
Thank you, Mike, for your incredibly kind note.  

I talked to Bill Claff as well and he graciously contributed the vast majority of the optical designs that I've included in the OpticExplorer database. 

Collaborating on tightly integrated code is usually pretty challenging but there are often loosely integrated approaches that can work.  I think what I'd like to do is to add to OpticExplorer the ability to read and write .roa files (or at least a subset of the format).   That way we can offer a loosely integrated workflow to sketch out the optical configuration in OpticExplorer and then perform analysis through RayOptics.  Currently, OpticExplorer is intended mainly for enthusiasts or for education so its analysis capacities are minimal.

I was curious about your thoughts about the RayOptics user base and applications.   Is RayOptics mainly intended for enthusiasts or do you see a population of more serious / professional users using RayOptics? I wonder how far away free / open source tools are from being able to offer an appealing alternative to commercial tools.  Is this something that you could imagine being possible in the not too distant future or do you think that free / open source tools are a very long way away from consideration for "serious" use?

If there's a .roa file format description that you can point me to that would be appreciated but otherwise, the structure of the .roa format appears to be mostly straightforward, at least to a rough approximation.  There are a number of items that are still in our queue (zoom lenses, mirrors etc.) so this might take a little time to get to but I think it would be a nice addition.  I can let you know when we have a working draft.

Thanks again for the encouragement!

**Reply by mjhoptics** [reply]
I can describe the `.roa` format in general, and write a top-level layout of the data  - a to-do for me. The .roa file is JSON formatted export of the OpticalModel in **ray-optics**. It's a nested collection of lists of objects, defined by dictionaries of their attributes. The object classnames and the modules they're defined in are written into the file, so they can be looked up in the documentation if there's a question (and hopefully there'll be documentation to find).

As far as my goals for **ray-optics**, I'm primarily driven by my interest in geometric optics and optical design and that I enjoy writing software. So I've looked into topics that have interested me in the past that I wasn't able to explore then. It's also been interesting to see what people wanted to do with **ray-optics**. Looking at patent designs is certainly one area, but there is also substantial interest from people trying to build up optical systems from catalog components. I been working on first order design tools, like the y-ybar diagram, with the idea of integrating component selection into that.

Professionals are always going to need access to the big commercial tools. Something like **ray-optics** might be interesting to the engineer who is comfortable with python to use of problems that aren't efficiently addressed by the big codes. Much of the interest I've received is from students, and that is somewhat by design.

To be considered as a possible design replacement, **ray-optics** would need optimization (at a minimum). I've worked on both the CODE V and LightTools optimizers and this is a lot of work. I don't have any immediate plans to look into this.

Regards, Mike

**Comment by mjhoptics** [comment]
Here's a quick attempt at laying out the .roa json file structure. Please ask questions!

# **ray-optics** .roa json file specification


```python
import json
import json_tricks
import pprint
```


```python
#from rayoptics.environment import *
from pathlib import Path
from rayoptics.gui import roafile
```

## Use Sasian Triplet.roa as an example


```python
filename = 'Sasian Triplet.roa'
```


```python
file_contents = roafile.preprocess_roa(Path.cwd() / filename, roafile.module_repl_050)
```

## The .roa file is *almost* pure json

An exception is how IEEE inf and nan are encoded.


```python
json_inputs = json.loads(file_contents)
```

## **ray-optics** python object encoding

**ray-optics** uses the **json_tricks** package for serializing python objects. The json_tricks package provides dump() and load() functions that augment the default python object encoding in json by including object metadata.

The default json encoding of a python object is to encode the python object's attributes as a json object. The json_tricks package encodes a python object as a json object consisting of the object metadata info ('__instance_type__') and the object 'attributes'. The dump() implementation recursively exports all of the object's attributes.

## Top level JSON object contains the optical model.

The attributes of the optical_model include the submodels (sequential, paraxial, element, etc)


```python
pprint.pprint(json_inputs, indent=1, sort_dicts=False, depth=3)
```

    {'optical_model': {'__instance_type__': ['rayoptics.optical.opticalmodel',
                                             'OpticalModel'],
                       'attributes': {'ro_version': '0.8.5',
                                      'radius_mode': True,
                                      'specsheet': {...},
                                      'system_spec': {...},
                                      'seq_model': {...},
                                      'optical_spec': {...},
                                      'parax_model': {...},
                                      'ele_model': {...},
                                      'part_tree': {...},
                                      'profile_dict': {...},
                                      'parts_dict': {...}}}}


## Optical Model attributes

The optical model data has 3 sections:

- the version and radius setting
- the data for each of the submodels
- lists of profiles and parts that are referenced in other submodels




```python
json_opm = json_inputs['optical_model']['attributes']
pprint.pprint(json_opm, sort_dicts=False, depth=3)
```

    {'ro_version': '0.8.5',
     'radius_mode': True,
     'specsheet': {'__instance_type__': ['rayoptics.parax.specsheet', 'SpecSheet'],
                   'attributes': {'conjugate_type': 'infinite',
                                  'imager': [...],
                                  'imager_inputs': {...},
                                  'frozen_imager_inputs': [...],
                                  'etendue_inputs': {...},
                                  'etendue_values': {...}}},
     'system_spec': {'__instance_type__': ['rayoptics.optical.opticalmodel',
                                           'SystemSpec'],
                     'attributes': {'title': '',
                                    'initials': '',
                                    '_dimensions': 'mm',
                                    'temperature': 20.0,
                                    'pressure': 760.0}},
     'seq_model': {'__instance_type__': ['rayoptics.seq.sequential',
                                         'SequentialModel'],
                   'attributes': {'ifcs': [...],
                                  'gaps': [...],
                                  'z_dir': [...],
                                  'do_apertures': True,
                                  'stop_surface': 3,
                                  'cur_surface': 6}},
     'optical_spec': {'__instance_type__': ['rayoptics.raytr.opticalspec',
                                            'OpticalSpecs'],
                      'attributes': {'spectral_region': {...},
                                     'pupil': {...},
                                     'field_of_view': {...},
                                     'defocus': {...}}},
     'parax_model': {'__instance_type__': ['rayoptics.parax.paraxialdesign',
                                           'ParaxialModel'],
                     'attributes': {'seq_mapping': None,
                                    'sys': [...],
                                    'ax': [...],
                                    'pr': [...],
                                    'ztwp': [...],
                                    'opt_inv': 2.274813964163764,
                                    'y_star': 6.249999999999999,
                                    'ybar_star': -3639702346.9129076}},
     'ele_model': {'__instance_type__': ['rayoptics.elem.elements', 'ElementModel'],
                   'attributes': {}},
     'part_tree': {'__instance_type__': ['rayoptics.elem.parttree', 'PartTree'],
                   'attributes': {'root_node': {...}}},
     'profile_dict': {'5803188112': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5803187680': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5803189312': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5811100288': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5811101728': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5811103888': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5811099472': {'__instance_type__': [...],
                                     'attributes': {...}},
                      '5803190560': {'__instance_type__': [...],
                                     'attributes': {...}}},
     'parts_dict': {'5803188880': {'__instance_type__': [...], 'attributes': {...}},
                    '5803190416': {'__instance_type__': [...], 'attributes': {...}},
                    '5803194112': {'__instance_type__': [...], 'attributes': {...}},
                    '5803188736': {'__instance_type__': [...], 'attributes': {...}},
                    '5803186768': {'__instance_type__': [...], 'attributes': {...}},
                    '5803192672': {'__instance_type__': [...], 'attributes': {...}},
                    '5803188496': {'__instance_type__': [...], 'attributes': {...}},
                    '5803186672': {'__instance_type__': [...], 'attributes': {...}},
                    '5811108928': {'__instance_type__': [...],
                                   'attributes': {...}}}}


## The Sequential Model

Only the `ifcs`, `gaps`, and `z_dir` arrays are needed to fully specify the model.


```python
json_sm = json_opm['seq_model']['attributes']
pprint.pprint(json_sm, sort_dicts=False, depth=2)
```

    {'ifcs': [{...}, {...}, {...}, {...}, {...}, {...}, {...}, {...}],
     'gaps': [{...}, {...}, {...}, {...}, {...}, {...}, {...}],
     'z_dir': [1, 1, 1, 1, 1, 1, 1],
     'do_apertures': True,
     'stop_surface': 3,
     'cur_surface': 6}


### Surface encoding example

Note the profile is encoded in a separate json object, json_opm['profile_dict']. The 'profile_id' attribute is a key to the actual profile, contained in the json_opm['profile_dict'].


```python
json_ifcs2 = json_sm['ifcs'][2]
pprint.pprint(json_ifcs2, sort_dicts=False, depth=3)
```

    {'__instance_type__': ['rayoptics.elem.surface', 'Surface'],
     'attributes': {'interact_mode': 'transmit',
                    'delta_n': -0.6910020663241183,
                    'decenter': None,
                    'max_aperture': 8.948204697323483,
                    'label': '',
                    'clear_apertures': [],
                    'edge_apertures': [],
                    'profile_id': '5803189312'}}


### Gap encoding example


```python
json_gaps3 = json_sm['gaps'][3]
pprint.pprint(json_gaps3, sort_dicts=False, depth=6)
```

    {'__instance_type__': ['rayoptics.seq.gap', 'Gap'],
     'attributes': {'thi': 0.975,
                    'medium': {'__instance_type__': ['opticalglass.schott',
                                                     'SchottGlass'],
                               'slots': {},
                               'attributes': {'gname': 'N-SF5',
                                              'coefs': {'__ndarray__': [1.52481889,
                                                                        0.187085527,
                                                                        1.42729015,
                                                                        0.011254756,
                                                                        0.0588995392,
                                                                        129.141675],
                                                        'dtype': 'float64',
                                                        'shape': [6]}}}}}


## Optical Spec


```python
json_osp = json_opm['optical_spec']['attributes']
pprint.pprint(json_osp, sort_dicts=False, depth=1)
```

    {'spectral_region': {...},
     'pupil': {...},
     'field_of_view': {...},
     'defocus': {...}}



```python
json_wvls = json_osp['spectral_region']['attributes']
pprint.pprint(json_wvls, sort_dicts=False, depth=2)
```

    {'wavelengths': [486.1327, 587.5618, 656.2725],
     'spectral_wts': [0.5, 1.0, 0.5],
     'render_colors': ['#268bd2', '#859900', '#dc322f'],
     'reference_wvl': 1,
     'coating_wvl': 550.0}



```python
json_pupil = json_osp['pupil']['attributes']
pprint.pprint(json_pupil, sort_dicts=False, depth=3)
```

    {'_key': ['aperture', 'object', 'epd'],
     'value': 12.5,
     'pupil_rays': [[0.0, 0.0], [1.0, 0.0], [-1.0, 0.0], [0.0, 1.0], [0.0, -1.0]],
     'ray_labels': ['00', '+X', '-X', '+Y', '-Y']}



```python
json_fov = json_osp['field_of_view']['attributes']
pprint.pprint(json_fov, sort_dicts=False, depth=2)
```

    {'key': ['field', 'object', 'angle'],
     'value': 20.0,
     'is_relative': True,
     'index_labels': ['axis', ' 0.71y', 'edge'],
     'index_label_type': 'auto',
     'fields': [{...}, {...}, {...}]}



```python
json_field2 = json_fov['fields'][2]
pprint.pprint(json_field2, sort_dicts=False, depth=2)
```

    {'__instance_type__': ['rayoptics.raytr.opticalspec', 'Field'],
     'attributes': {'x': 0.0,
                    'y': 1.0,
                    'vux': 0.0,
                    'vuy': 0.0,
                    'vlx': 0.0,
                    'vly': 0.0,
                    'wt': 1.0,
                    'aim_pt': {...}}}



```python
json_defocus = json_osp['defocus']['attributes']
pprint.pprint(json_defocus, sort_dicts=False, depth=1)
```

    {'focus_shift': 0.0, 'defocus_range': 0.0}



```python
json_pm = json_opm['parax_model']['attributes']
pprint.pprint(json_pm, sort_dicts=False, depth=3)
```

    {'seq_mapping': None,
     'sys': [[0.0, 10000000000.0, 1.0, 'dummy'],
             [0.029140221242530184,
              2.856885923564585,
              1.6910020663241183,
              'transmit'],
             [-9.425384275234017e-05, 5.86, 1.0, 'transmit'],
             [-0.027506830028392514,
              0.5828874868684721,
              1.6727070351743674,
              'transmit'],
             [-0.03072282769338543, 4.822, 1.0, 'transmit'],
             [0.007964615386577972,
              1.8491993961884612,
              1.6910020663241183,
              'transmit'],
             [0.03371695730129101, 41.2365, 1.0, 'transmit'],
             [0.0, 0.0, 1.0, 'dummy']],
     'ax': [[0.0, 6.249999992700494e-10],
            [6.249999992700494, -0.18212638192810443],
            [5.729685695860345, -0.1815863370335065],
            [4.665589760843997, -0.05325075249976219],
            [4.634550563545556, 0.08913574590033027],
            [5.0643631302769485, 0.0488000413897083],
            [5.154604137348769, -0.12499752621433813],
            [0.000143647611214881, -0.12499752621433813]],
     'pr': [[-3639702346.9129076, 0.36397023426620234],
            [-4.250884056091309, 0.4878419361370471],
            [-2.8571752958168855, 0.48757263638599935],
            [3.5340507098524654e-07, 0.4875726461070525],
            [0.28420034776022174, 0.49630408442169],
            [2.677378642841611, 0.47497979328721845],
            [3.5557109897900556, 0.355092037668736],
            [18.198463801116887, 0.355092037668736]],
     'ztwp': [[{...}, {...}, {...}, {...}, {...}, {...}, {...}],
              [22748139641.637638,
               6.49888399294761,
               13.330409829999658,
               1.325960594664723,
               10.96915293519767,
               4.206584608972712,
               93.80536603323905],
              [{...}, {...}, {...}, {...}, {...}, {...}, {...}],
              [0.0,
               0.012809935977882181,
               -4.1433648745421025e-05,
               -0.012091903101405569,
               -0.013505644055899463,
               0.0035012161486822134,
               0.01482185261408204]],
     'opt_inv': 2.274813964163764,
     'y_star': 6.249999999999999,
     'ybar_star': -3639702346.9129076}



```python
json_profiles = json_opm['profile_dict']
pprint.pprint(json_profiles, sort_dicts=False, depth=3)
```

    {'5803188112': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.0}},
     '5803187680': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.042170961076202926}},
     '5803189312': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.00013640168003221264}},
     '5811100288': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': -0.04088976120379457}},
     '5811101728': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.04567044208987943}},
     '5811103888': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.011526181721781025}},
     '5811099472': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': -0.04879429301948844}},
     '5803190560': {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
                    'attributes': {'cv': 0.0}}}



```python
json_profile2 = json_profiles['5803189312']
pprint.pprint(json_profile2, sort_dicts=False, depth=3)
```

    {'__instance_type__': ['rayoptics.elem.profiles', 'Spherical'],
     'attributes': {'cv': 0.00013640168003221264}}



```python
json_em = json_opm['ele_model']['attributes']
pprint.pprint(json_em, sort_dicts=False, depth=2)
```

    {}



```python
json_pt = json_opm['part_tree']['attributes']
pprint.pprint(json_pt, sort_dicts=False, depth=3)
```

    {'root_node': {'id_key': '5803186624',
                   'tag': '#group#root',
                   'name': 'root',
                   'children': [{...},
                                {...},
                                {...},
                                {...},
                                {...},
                                {...},
                                {...},
                                {...},
                                {...}]}}



```python
json_parts = json_opm['parts_dict']
pprint.pprint(json_parts, sort_dicts=False, depth=2)
```

    {'5803188880': {'__instance_type__': [...], 'attributes': {...}},
     '5803190416': {'__instance_type__': [...], 'attributes': {...}},
     '5803194112': {'__instance_type__': [...], 'attributes': {...}},
     '5803188736': {'__instance_type__': [...], 'attributes': {...}},
     '5803186768': {'__instance_type__': [...], 'attributes': {...}},
     '5803192672': {'__instance_type__': [...], 'attributes': {...}},
     '5803188496': {'__instance_type__': [...], 'attributes': {...}},
     '5803186672': {'__instance_type__': [...], 'attributes': {...}},
     '5811108928': {'__instance_type__': [...], 'attributes': {...}}}



```python
json_part2 = json_parts['5803192672']
pprint.pprint(json_part2, sort_dicts=False, depth=6)
```

    {'__instance_type__': ['rayoptics.elem.elements', 'Element'],
     'attributes': {'label': 'E2',
                    'tfrm': [{'__ndarray__': [[1.0, 0.0, 0.0],
                                              [0.0, 1.0, 0.0],
                                              [0.0, 0.0, 1.0]],
                              'dtype': 'float64',
                              'shape': [3, 3],
                              'Corder': True},
                             {'__ndarray__': [0.0, 0.0, 10.691],
                              'dtype': 'float64',
                              'shape': [3]}],
                    's1_indx': 3,
                    's2_indx': 4,
                    'medium_name': 'N-SF5',
                    '_sd': 4.791917244000941,
                    'hole_sd': None,
                    'flat1': None,
                    'flat2': None,
                    'do_flat1': 'if concave',
                    'do_flat2': 'if concave',
                    'edge_extent': [-4.791917244000941, 4.791917244000941],
                    'profile1_id': '5811100288',
                    'profile2_id': '5811101728'}}

**Reply by amegahed** [reply]
Thank you Michael.   I implemented an initial draft of an .roa importer:
https://opticexplorer.sharedigm.com/#search?query=Sasian

It only handles spherical and planar surfaces and doesn't do paraxial lenses.  It's only pulling materials from the Schott catalog.  I'm sure there are many shortcomings.   The real value is probably not an importer, though, but an exporter, since the expected workflow is probably to model in OpticExplorer and analyze in RayOptics.

**Reply by mjhoptics** [reply]
That looks encouraging. I’d look at the Optical Spec data so that you can set up the fields, etc., for the system specification.
I’m on the road right now and will be at the IODC in Quebec next week. Please forgive me if I don’t respond promptly.
Thanks!
Mike Hayford
